docs/models/advanced/extended-blockstates.md

```diff
@@ -0,0 +1,19 @@
+Extended Blockstates
+====================
+
+Extended blockstates provide a way for blocks to pass arbitrary data to their models. Ordinary blockstates and properties occupy only a fixed set of possible states, while extended states can form infinite sets. This accomplished through the use of unlisted properties (`IUnlistedProperty<V>`), extended block states (`IExtendedBlockState`), and extended block state containers (`ExtendedBlockState`).
+
+Ordinary block state containers (`BlockStateContainer`) take a set of listed properties (`IProperty<V>`) that define all possible values for themselves and use those to create all possible combinatons of values. These combinations are then turned into `IBlockState`s and stored in the finite set of all possible states. The properties are called "listed" as they appear on the F3 debug screen, all the way to the right.
+
+Extended block state containers take a set of listed properties, and also a set of *unlisted* properties. Unlisted properties do not define all of their possible values, but instead provide a predicate for checking if a value is valid. The listed properties are again used to create the set of states, this time `IExtendedBlockState`s, but the unlisted properties remain unset. The unlisted properties are only set when `IExtendedBlockState::withProperty` is called to do so. This allows unlisted properties to have any type, even other `IBlockState`s.
```

Comment:
> An IUnlistedProperty still only has *1* value in the IExtendedBlockState.

I never implied the opposite, and I'm confused what part of what I wrote made you think that.

> `isValid` does not have to do with not having fixed values, it's just a safety check used in `withProperty`.

`withProperty` is rejecting values that fail the `isValid` check with this message:

```java
throw new IllegalArgumentException("Cannot set unlisted property " + property + " to " + value + " on block " + Block.REGISTRY.getNameForObject(getBlock()) + ", it is not an allowed value");
```

which implies, to me, that it's a predicate that checks whether a value is allowed for the property. I'm trying to contrast this against the fixed set of values `IProperty` uses.